lib/std/testing/failing_allocator.zig

```diff
@@ -39,43 +39,49 @@ pub const FailingAllocator = struct {
             .allocations = 0,
             .deallocations = 0,
             .allocator = mem.Allocator{
-                .reallocFn = realloc,
-                .shrinkFn = shrink,
+                .allocFn = alloc,
+                .capacityFn = capacity,
+                .resizeFn = resize,
             },
         };
     }
 
-    fn realloc(allocator: *mem.Allocator, old_mem: []u8, old_align: u29, new_size: usize, new_align: u29) ![]u8 {
-        const self = @fieldParentPtr(FailingAllocator, "allocator", allocator);
+    fn alloc(allocator: *std.mem.Allocator, len: usize, alignment: u29) error{OutOfMemory}![*]u8 {
+        const self = @fieldParentPtr(@This(), "allocator", allocator);
         if (self.index == self.fail_index) {
             return error.OutOfMemory;
         }
-        const result = try self.internal_allocator.reallocFn(
-            self.internal_allocator,
-            old_mem,
-            old_align,
-            new_size,
-            new_align,
-        );
-        if (new_size < old_mem.len) {
-            self.freed_bytes += old_mem.len - new_size;
-            if (new_size == 0)
-                self.deallocations += 1;
-        } else if (new_size > old_mem.len) {
-            self.allocated_bytes += new_size - old_mem.len;
-            if (old_mem.len == 0)
-                self.allocations += 1;
-        }
+        const result = try self.internal_allocator.callAllocFn(len, alignment);
+        const result_capacity = self.internal_allocator.callCapacityFn(result[0..len]);
```

Comment:
I'm confused why you would need the capacity function here. Wouldn't the capacity be `result.len`?

Reply:
Copied from the description:
> Note that there was another proposal to return the full capacity of an allocation in the allocFn function. This created a problem with some of the current behavior implemented by the standard library. For example, the failing_allocator.zig keeps track of all the memory that has been allocated and freed. Because the resizeFn function accepts either the requested size, or the full capacity, a "wrapping allocator" doesn't know whether the buffers it receives have the requested length or the full capacity, only the underlying allocator knows that. So it's unable to know how much memory has actually been allocated and freed. By adding the capacityFn to the allocator interface, any code that has an allocation can query the full capacity from the allocator at any time.

Reply:
Resolving since we no longer need a capacity function